traffic_monitor/health/cache.go

```diff
@@ -171,45 +172,100 @@ func EvalCache(result cache.ResultInfo, resultStats *threadsafe.ResultStatValHis
 		}
 
 		if !inThreshold(threshold, resultStatNum) {
-			return false, eventDesc(status, exceedsThresholdMsg(stat, threshold, resultStatNum)), stat
+			return false, result.UsingIPv4, eventDesc(status, exceedsThresholdMsg(stat, threshold, resultStatNum)), stat
 		}
 	}
 
-	return avail, eventDescVal, eventMsg
+	return avail, result.UsingIPv4, eventDescVal, eventMsg
 }
 
 // CalcAvailabilityWithStats calculates the availability of each cache in results.
 // statResultHistory may be nil, in which case stats won't be used to calculate availability.
-func CalcAvailability(results []cache.Result, pollerName string, statResultHistory *threadsafe.ResultStatHistory, mc tc.TrafficMonitorConfigMap, toData todata.TOData, localCacheStatusThreadsafe threadsafe.CacheAvailableStatus, localStates peer.CRStatesThreadsafe, events ThreadsafeEvents) {
+func CalcAvailability(results []cache.Result, pollerName string, statResultHistory *threadsafe.ResultStatHistory, mc tc.TrafficMonitorConfigMap, toData todata.TOData, localCacheStatusThreadsafe threadsafe.CacheAvailableStatus, localStates peer.CRStatesThreadsafe, events ThreadsafeEvents, protocol config.PollingProtocol) {
 	localCacheStatuses := localCacheStatusThreadsafe.Get().Copy()
 	statResults := (*threadsafe.ResultStatValHistory)(nil)
+	processAvailableTuple := func(tuple cache.AvailableTuple, serverInfo tc.TrafficServer) bool {
+		switch protocol {
+		case config.IPv4Only:
+			return tuple.IPv4
+		case config.IPv6Only:
+			return tuple.IPv6
+		case config.Both:
+			// only report availability based on defined IP addresses
+			if serverInfo.IP == "" {
+				return tuple.IPv6
+			} else if serverInfo.IP6 == "" {
+				return tuple.IPv4
+			}
+			// if both IP addresses are defined then report availability based on both
+			return tuple.IPv4 || tuple.IPv6
```

Comment:
This returns that the cache is healthy if _either_ IPv4 _or_ IPv6 polling reported available? Is that intentional?
If so, I think we should document that. I didn't see anything in the docs indicating that "both" is really "either."

It'd be nice if it were a config option, for "both" and "either." But I don't want to ask you to do more work, this works as-is, as long as we document the behavior. We can always add "either" later.

Reply:
yea i have it doing either because then TR can find all of the healthy caches even if only one is available. for example if IPv4 has an error but IPv6 is still reporting healthy, then TR can still direct traffic to the IPv6 address.  I like the idea of having it as a config option though so people can choose for themselves